tests/unit/providers/test_vcluster_provider.py

```
@@ -0,0 +1,511 @@
+    assert "https://127.0.0.1:31234" in content
+
+
+def test_vcluster_ensure_cluster_credentials_missing_outputs_raises() -> None:
+    with pytest.raises(ConfigError, match="must be a non-empty string"):
+        VClusterProvider().ensure_cluster_credentials("c", "local", {}, outputs=None)
+
+
+def test_vcluster_ensure_cluster_credentials_refuses_default_kubeconfig(
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    with pytest.raises(ConfigError, match="Refusing to overwrite ~/.kube/config"):
```

Ruff (0.16.1)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)